internal/resources/providers/gcplib/inventory/provider.go

```diff
@@ -91,242 +75,327 @@ type dnsPolicyFields struct {
+	defer close(out)
+
+	projectsCh := make(chan *assetpb.Asset)
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_RESOURCE, []string{CrmProjectAssetType}, projectsCh)
+
+	for project := range projectsCh {
+		logsResourceCh := make(chan *assetpb.Asset)
+		alertsResourceCh := make(chan *assetpb.Asset)
+		logsAssetCh := make(chan *ExtendedGcpAsset)
+		alertsAssetCh := make(chan *ExtendedGcpAsset)
+
+		go p.getAllAssets(ctx, project.Ancestors[0], assetpb.ContentType_RESOURCE, []string{MonitoringLogMetricAssetType}, logsResourceCh)
+		go p.getAllAssets(ctx, project.Ancestors[0], assetpb.ContentType_RESOURCE, []string{MonitoringAlertPolicyAssetType}, alertsResourceCh)
+		go p.enrichAssets(ctx, logsResourceCh, logsAssetCh)
+		go p.enrichAssets(ctx, alertsResourceCh, alertsAssetCh)
+
+		var logAssets, alertAssets []*ExtendedGcpAsset
+		var wg sync.WaitGroup
+
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			logAssets = collect(logsAssetCh)
+			p.log.Debugf("Listed %d log metrics for %v\n", len(logAssets), project.Name)
+		}()
+		go func() {
+			defer wg.Done()
+			alertAssets = collect(alertsAssetCh)
+			p.log.Debugf("Listed %d alert policies for %v\n", len(alertAssets), project.Name)
+		}()
+		wg.Wait()
+		out <- &MonitoringAsset{
+			LogMetrics:   logAssets,
+			Alerts:       alertAssets,
+			CloudAccount: logAssets[0].CloudAccount,
+		}
 	}
+}
 
-	alertPolicies, err := p.ListAllAssetTypesByName(ctx, monitoringAssetTypes["AlertPolicy"])
-	if err != nil {
-		return nil, err
-	}
+func (p *Provider) ListProjectsAncestorsPolicies(ctx context.Context, out chan<- *ProjectPoliciesAsset) {
+	defer close(out)
+
+	prjCh := make(chan *assetpb.Asset)
+	prjMetadataCh := make(chan *ExtendedGcpAsset)
+	policiesCache := &sync.Map{}
 
-	typeGenerator := func(assets []*ExtendedGcpAsset, projectId, projectName, orgId, orgName string) *MonitoringAsset {
-		return &MonitoringAsset{
-			LogMetrics: getAssetsByType(assets, MonitoringLogMetricAssetType),
-			Alerts:     getAssetsByType(assets, MonitoringAlertPolicyAssetType),
-			CloudAccount: &fetching.CloudAccountMetadata{
-				AccountId:        projectId,
-				AccountName:      projectName,
-				OrganisationId:   orgId,
-				OrganizationName: orgName,
-			},
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_IAM_POLICY, []string{CrmProjectAssetType}, prjCh)
+	go p.enrichAssets(ctx, prjCh, prjMetadataCh)
+
+	for asset := range prjMetadataCh {
+		out <- &ProjectPoliciesAsset{
+			CloudAccount: asset.CloudAccount,
+			Policies:     append([]*ExtendedGcpAsset{asset}, p.getAssetAncestorsPolicies(ctx, asset.Ancestors[1:], policiesCache)...),
 		}
 	}
+}
 
-	var assets []*ExtendedGcpAsset
-	assets = append(append(assets, logMetrics...), alertPolicies...)
-	monitoringAssets := getAssetsByProject[MonitoringAsset](assets, p.log, typeGenerator)
+func (p *Provider) ListProjectAssets(ctx context.Context, assetTypes []string, out chan<- *ProjectAssets) {
```

Comment:
this function now handles the same thing `ListLoggingAssets` and `ListServiceUsageAssets` used to. they both used to fetch both resource and policy for some asset types, group them all by project and send a slice with each group.

it now does the same thing, but starts by fetching projects and then for each project fetches only the resource of the given asset types and the sends it out. the difference is that we don't wait for all assets from all projects, plus we don't fetch policies as they weren't used by the relevant CIS rules.